File: DFS/depthfirstsearch.py
# Take in 2d array in the structure
# node1:
# node2:
# distance:
# return shortest distance from start to end
def node_count(arr1, arr2):
    max_val = arr1[0]
    for i in range(len(arr1)):
        if arr1[i] > max_val:
            max_val = arr1[i]
    print("New max:", max_val)
    for i in range(len(arr2)):
        if arr2[i] > max_val:
            max_val = arr2[i]
    print("Final max:", max_val)
    return max_val

def link(start, end, graph, visited=None):
    if visited is None:
        visited = set()

    if start == end:
        return 0

    visited.add(start)
    shortest = float('inf')

    for i in range(len(graph[start])):
        if graph[start][i] != 0 and i not in visited:
            dist = graph[start][i]
            result = link(i, end, graph, visited.copy())
            if result is not False and isinstance(result, (int, float)) and result != float('inf'):
                shortest = min(shortest, dist + result)
                print(f"Path found: {start} → {i} → {end} = {dist} + {result} = {shortest}")

    if shortest == float('inf'):
        return False
    return shortest

def dfs(node1arr, node2arr, distarr):
    highest = node_count(node1arr, node2arr)
    size = highest + 1
    graph = [[0 for _ in range(size)] for _ in range(size)]

    for i in range(len(node1arr)):
        from_node = node1arr[i]
        to_node = node2arr[i]
        dist = distarr[i]
        graph[from_node][to_node] = dist

    shortest = link(1, highest, graph)
    print(f"Shortest distance from 1 to {highest}:", shortest)
    return shortest

input_string = input("Enter from array (comma or space separated): ")
arr_from = [int(x) for x in input_string.replace(',', ' ').split()]

input_string = input("Enter to array (comma or space separated): ")
arr_to = [int(x) for x in input_string.replace(',', ' ').split()]

input_string = input("Enter dist array (comma or space separated): ")
arr_dist = [int(x) for x in input_string.replace(',', ' ').split()]

print("Result:", dfs(arr_from, arr_to, arr_dist))

File: DFS/depthfirstsearch.js
// Take in 2d array in the structure
// node1:
// node2:
// distance:
// return shortest distance from start to end
function nodeCount(arr1,arr2)
{
    let max = arr1[0];
    for(let i=0; i<arr1.length; i++)
    {
        if(arr1[i] > max) max = arr1[i];
    }
    console.log("New max: ", max);
    for(let i=0; i<arr2.length; i++)
    {
        if(arr2[i] > max) max = arr2[i];
    }
    console.log("Final max: ",max)    
    return max;
}

function link(start,end,map,visited = new Set())
{
    if(start == end) return 0;
    visited.add(start);
    let shortest = Infinity; 
    
    for(let i=0;i<map[start].length;i++)
    {
        if(map[start][i] !== 0 && !visited.has(i))
        {
            let dist = map[start][i];
            let result = link(i, end, map, new Set(visited));
            if (typeof result == "number" && result !== Infinity) 
            {
                shortest = Math.min(shortest, dist + result);
                console.log(`Path found: ${start} → ${i} → ${end} = ${dist} + ${result} = ${shortest}`);
            }
        }
    }
    if(shortest == Infinity) return false;
    return shortest;
}

function dfs(node1arr,node2arr,distarr)
{
    let highest = nodeCount(node1arr,node2arr);
    let count = node1arr.length;
    let map = new Array(highest).fill(null).map(() => new Array(highest).fill(0))
    for(let i=0; i<count; i++)
    {
        let from = node1arr[i];
        let to = node2arr[i];
        let dist = distarr[i];
        map[from][to] = dist;
    }
    let shortest = link(1,highest,map);
    console.log("Shortest distance from 1 to",highest,":")
    return shortest;
}

// dfs([1,2,3,1],[2,3,4,4],[10,20,39,20]);

const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout,
});

readline.question('Enter from array (comma or space separated): ', (input) => 
{
    let arrfrom = input.split(/[\s,]+/).map(Number);
    readline.question('Enter to array (comma or space separated): ', (input) => 
    {
        let arrto = input.split(/[\s,]+/).map(Number);
        readline.question('Enter dist array (comma or space separated): ', (input) => 
        {
            let distarr = input.split(/[\s,]+/).map(Number);
            dfs(arrfrom,arrto,distarr);
            readline.close();
        });
    });
});
